### DISCUSSION TITLE
Change to the beets UI

### DISCUSSION THREAD

**Discussion by davidswarbrick** [beetbox/beets#4870] [Polls]
Hi all,

Over the past 8 (:cold_sweat:) years there's been some suggestions for a new beets UI, with the code detailed in PR #3721 .

It looks like this:
![8d944ca2d70ae6c63cf77d2a940b46c391069fcd](https://github.com/beetbox/beets/assets/16927315/19213681-578f-4696-b264-9c981ab26993)

There's a [discourse discussion](https://discourse.beets.io/t/a-reasonably-new-commandline-ui-for-beets/1831/5) about it, in which I detail some instructions on how to setup and test the branch. However, in order to ensure those who prefer GitHub Discussions aren't left out, I thought I'd make a poll here to gauge community interest in the changes.
Please feel free to vote if you're unable to test, although it would be extra helpful if you can try out my changes and report back any bugs or preferences for things you'd like to see altered.

I'm a huge fan of beets and its community, I hope you enjoy these changes as much as I have when using them myself :smile:

**Comment by JOJ0** [comment]
If you've installed beets from the official git repo already, here's how to quickly checkout the feature branch within that repo:

```
git remote add davidswarbrick https://github.com/davidswarbrick/beets.git
git fetch -v davidswarbrick
git checkout --track davidswarbrick/ui
```

HTH

**Reply by arsaboo** [reply]
@JOJ0 I normally install/upgrade beets using `pip install --upgrade --force-reinstall --no-deps git+https://github.com/beetbox/beets.git`.

How do you recommend I checkout this branch?

**Reply by JOJ0** [reply]
The syntax for adding a branch with your command is attaching `@branch` to the repo uri.

I think 

https://github.com/davidswarbrick/beets.git@ui

**Reply by JOJ0** [reply]
@arsaboo did that work?

**Comment by JOJ0** [comment]
Another option, and what I personally do, is cherry picking commits into my "personal beets version". In this case we only have 2 commits, so that's straight forward (providing the remote and its data is known already - see my prev post):

```
git checkout my_personal_dev_branch
git cherry-pick 23f8e0df 9f6236bb
```

**Comment by Emporea** [comment]
What about beetarr?

**Reply by JOJ0** [reply]
Hi, what is / or does beetarr mean @Emporea?

**Reply by Emporea** [reply]
I mean this:

https://github.com/sentriz/betanin

**Comment by JOJ0** [comment]
Thoughts about colors thread.

**Reply by arsaboo** [reply]
The colors look different (and less readable) on my end:

<img width="447" alt="image" src="https://github.com/beetbox/beets/assets/18319734/220b3915-d8da-4b3d-be50-ef2e5e2c3af0">

@davidswarbrick do you have any special config for UI colors?

**Reply by JOJ0** [reply]
From David's screenshot it looks like he might use the same terminal color scheme I'm using, which is called "Tomorrow Night Eighties"

In your case @arsaboo you could either play with different themes in your terminal oor just change that particular setting. There is new possibilities in the beets config now. Just a moment...

**Reply by JOJ0** [reply]
Not sure which one is for the "white in blue" text but see the diff of config_default.yaml: https://github.com/beetbox/beets/commit/23f8e0df1bdb149224cc33d3af16e91dc20f0efc

Also: What OS and term app u using?

**Reply by arsaboo** [reply]
Beets is installed on a Ubuntu machine and I am accessing it using MTputty.

**Reply by JOJ0** [reply]
So you are accessing it from a Windows machine using Putty, right @arsaboo? Does your Putty version (https://ttyplus.com/multi-tabbed-putty/) have the same color settings / color themes like the regular Putty? Is there any color themes to choose from?

So what I'm thinking is: Maybe the default colors are a little "too advanced / not good" for Windows/Putty. @davidswarbrick I  think we should ship this new gui with failsafe color settings (even though after 5 minutes I started to love what you chose for the defaults! It just looks awesome and also the blue banner-like path of the folder to be imported really helps to see where one folder starts and the other stopped, kind of like a big fat "border line")

Anyway I suggest for the latter we could just remove the blue background color and just leave the text white.

What also could be problematic is blue text in general (the choices) on Windows and "bad" default themes.

But I don't want to loose the settings you suggest, maybe we could add to the docs on how to get them back in "good" terminal apps with proper default themes.

Some randomish thoughts......help me sort :-)

**Reply by arsaboo** [reply]
Mtputty just uses Putty internally...so I am guessing we should be able to use the basic themes. Also, agree that we should ship this with a basic failsafe color setting to avoid such issues.

**Reply by JOJ0** [reply]
Please, if you have played around with these "hard to read" color-settings on your end, post here which one's to change to make it better readable. I guess it must be only one or two variables that have to be changed in config_default.yaml. When we have it documented in here, David or me could easily integrate it into the PR.

**Reply by davidswarbrick** [reply]
I've managed to recreate @arsaboo's setup (i.e. PuTTY defaults). I found the best balance between readability & design seems to be this combination:
![image](https://github.com/beetbox/beets/assets/16927315/e0a263a9-c04f-4b3f-b780-17e2b7c40bf3)
The diff to recreate is:
```diff
--- a/beets/config_default.yaml
+++ b/beets/config_default.yaml
@@ -102,11 +102,11 @@ ui:
         text_highlight: ['bold', 'red']
         text_highlight_minor: ['white']
         action_default: ['bold', 'cyan']
-        action: ['bold', 'blue']
+        action: ['bold', 'cyan']
         # New Colors
         text: ['normal']
         text_faint: ['faint']
-        import_path: ['bold', 'inverse', 'blue']
+        import_path: ['bold', 'blue']
         import_path_items: ['bold', 'blue']
         added:   ['green']
         removed: ['red']
@@ -117,7 +117,7 @@ ui:
         text_diff_added:   ['bold', 'red']
         text_diff_removed: ['bold', 'red']
         text_diff_changed: ['bold', 'red']
-        action_description: ['blue']
+        action_description: ['white']
     import:
         indentation:
             match_header: 2
```
 Window 10's default cmd.exe looks much easier on the eye than PuTTY (and hence is now my usual default on Windows systems), with the original config:
![Screenshot from 2023-09-02 14-58-43](https://github.com/beetbox/beets/assets/16927315/7fc6b7f0-b459-494f-96df-5e716e4ce7b7)

**Reply by JOJ0** [reply]
I actually wanted to say: Screw this, we can't support any terminal app with bad colorscheme support (Sorry putty I didn't want to hurt your feelings, but there is Windows Terminal now and duh it does a better job than you do ;)))) hahaha);
but since you changed this quickly David, I'm very fine with it and certainly stay with my opinion that going fail-safe is probably the best choice for new beets users. 

Nevertheless what I realise now is that we should add some docs https://beets.readthedocs.io/en/latest/reference/config.html#colors, well, we _could_ add something new. Actually the docs are not outdated, those options described are still valid with the new UI. We could even make it quick and point to the "self-descriptive names in config_default.yaml" - that should be enough for colors IMHO.

The import options block though requires some docs I think: https://github.com/beetbox/beets/pull/3721/files#diff-608081b9e86676fedcbc02b402def894b85f87a8c4e6f0495a87bc2119cbb8bdR121-R126

Anyway @davidswarbrick I want to hit that merge button soon! I would also like to help with docs and will try during the next week, but can't promise since I _should_ be working on something else actually which I kind of postpone constantly......haha

**Reply by RollingStar** [reply]
Because of this ui branch I souped up my putty config and it's much better. https://www.linux.com/training-tutorials/improving-putty-settings-windows/

**Reply by JOJ0** [reply]
Especially this part. OMG! Finally someone writing this down... 😆
![image](https://github.com/beetbox/beets/assets/2733783/70f52f73-b18d-4cf6-b3b5-e5702cb10a8a)

**Comment by JOJ0** [comment]
As @arsaboo suggested, there is another simple way of installing beets from this development version:

```
pip install --upgrade --force-reinstall --no-deps git+https://github.com/davidswarbrick/beets.git@ui
```

To revert your installation to the official git master version again, do

```
pip install --upgrade --force-reinstall --no-deps git+https://github.com/beetbox/beets.git
```

**Reply by arsaboo** [reply]
Yes, that is how I am currently installing this dev version.

**Comment by RollingStar** [comment]
Some screenshots:

Advanced putty: looks excellent! [I recommend changing putty defaults if yours doesn't look like this.](https://www.linux.com/training-tutorials/improving-putty-settings-windows/)

![image](https://github.com/beetbox/beets/assets/4810902/a86e7f04-eb96-4505-963b-0bc51fddefa0)

![image](https://github.com/beetbox/beets/assets/4810902/6808a4a6-aee6-4b28-b137-5debed7728d7)

![image](https://github.com/beetbox/beets/assets/4810902/9010dd2a-8429-49da-a50d-6e7c7c3e3323)

![image](https://github.com/beetbox/beets/assets/4810902/71145d82-4a14-4885-ae57-1fc29632465e)

![image](https://github.com/beetbox/beets/assets/4810902/33e803ac-19a7-46f1-8966-2ba1040df9a0)

Basic putty - looks pretty good! 

![image](https://github.com/beetbox/beets/assets/4810902/827c6adc-0c3e-426d-8065-97d127272b90)

![image](https://github.com/beetbox/beets/assets/4810902/76538d82-c42f-4fff-8691-bfa141e5a270)

![image](https://github.com/beetbox/beets/assets/4810902/c3605fac-aa83-4c48-bf13-d6382979965a)

![image](https://github.com/beetbox/beets/assets/4810902/cdc70a9b-dbe3-4285-948e-b9810adc56ed)

![image](https://github.com/beetbox/beets/assets/4810902/8e75d390-5685-4054-8873-2e244de15205)

For completeness, here's how it looks in -vv mode - not ideal, but out of scope here.

![image](https://github.com/beetbox/beets/assets/4810902/dc09f1d2-e364-4f15-bd69-70a2713f2184)
